weaveworks/eksctl
# Add correct Ubuntu bootstrap script

### Description

eksctl is injecting the wrong bootstrap script into ubuntu images, causing incorrect behavior of worker nodes running ubuntu EKS images. This PR adds the correct bootstrap script to the nodegroup assets. The following text describes testing with a k8s 1.12 configuration, but the bootstrap script will work for all Ubuntu EKS images built for k8s 1.11 and higher.

Please note that this PR represents a workaround only. The eksctl tool should not need to inject custom bootstrapping behavior into an official Ubuntu EKS image, as it comes with a bootstrap script baked in (the same one added by this PR). Further discussion on this topic is forthcoming.

I've tested these changes manually (description follows) with success. This is great, considering these tests fail when using the current stable version of eksctl, following the instructions on the AWS EKS getting started guide. However, `make build` produces a `pkg/nodebootstrap/assets.go` file different from the one here (in ways other than simply the addition of code artifacts as a result of the new bootstrap script), and I'm not quite sure why. I cannot be sure I am reproducing a production build of eksctl, so my test successes come with that caveat.

Testing:

1. Run the three `make` commands listed on the contributor readme for this project
2. Launch an EKS cluster with the following command:
```
./eksctl create cluster \
  --name <cluster name> \
  --version 1.12 \
  --nodegroup-name test-workers \
  --node-type m5.large \
  --nodes 10 \
  --nodes-min 10 \
  --nodes-max 10 \
  --node-ami ami-0551b6dc0b4079e1d \ # this is the official Ubuntu 18.04 ami for k8s 1.12 in us-west-2: https://cloud-images.ubuntu.com/aws-eks/
  --node-ami-family Ubuntu1804 \
  --region us-west-2 \
  --ssh-access \
  --ssh-public-key <AWS .pem key>
```
3. Wait for cluster launch to complete
4. Install kubectl
`sudo snap install kubectl --classic --channel 1.12/stable`
5. Verify connectivity
`kubectl get all --all-namespaces`
6. Install/Config/Run the tests (should take < 2 hours):
```
sudo snap install kubernetes-test-eks --classic --channel 1.12.9/stable
export SERVER=$(cat $HOME/.kube/config | grep server | sed 's/    server: //') ## (Note that this only works if you have only one cluster listed in your config)
export FOCUS="\\[Conformance\\]"
export SKIP="\\[Flaky\\]|\\[Serial\\]"
export ACTION_JUNIT=$HOME/eks/e2e
export ACTION_LOG=$HOME/eks/e2e.log
export KUBECONFIG=$HOME/.kube/config
export PARALLELISM=10 ## ← how many e2e threads to run
mkdir -p $HOME/eks/e2e
GINKGO_ARGS="-nodes=$PARALLELISM" kubernetes-test-eks.e2e \
  -kubeconfig $HOME/.kube/config \
  -host $SERVER \
  -ginkgo.focus "$FOCUS" \
  -ginkgo.skip "$SKIP" \
  -report-dir $ACTION_JUNIT 2>&1 | tee -a $ACTION_LOG
```
Success: 171 Passed, 7 Failed, 0 Pending, 1834 Skipped
Failure: 10 or more tests fail

Testing using the current stable eksctl package results in failure due to, perhaps among other problems, an apparent misconfiguration of CA configs, which results in worker node communication failures.

### Checklist
- [ ] Added/modified documentation as required (such as the `README.md`, and `examples` directory)
